### ISSUE TITLE
Windows undocks when changing tabs

### ISSUE THREAD

**Issue by hka** [ocornut/imgui#6505]
**Version/Branch of Dear ImGui:**

Version: 1.89.7 WIP
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: custom
Compiler: gcc 10.2.1
Operating System: Debian 11

**My Issue/Question:**

Windows get undocked when changing tab even when I call DockSpace with ImGuiDockNodeFlags_KeepAliveOnly for the hidden tabs. I have looked at issues #2599 and #2720 and I think I do as it says in those issues. Am I doing something wrong, or what should I do to make a DockSpace stay put inside a tab? I have mostly picked parts from imgui_demo.cpp

**Screenshots/Video**
A build off my code: 
https://hka.github.io/test-imgui/
press the plus to create a second tab, press "Menu/New workspace" to get a window to dock, dock it and change tab.

**Standalone, minimal, complete and verifiable example:** 
Not complete really but I guess this is the part where I do something wrong, for the rest of the code see 
https://github.com/hka/test-imgui/blob/master/src/simple.cpp
not sure how make a more minimal example, very new to imgui.
```
  if (ImGui::BeginTabBar("MyTabBar", tab_bar_flags))
  {
    if (ImGui::TabItemButton("+", ImGuiTabItemFlags_Trailing | ImGuiTabItemFlags_NoTooltip))
    {
      active_tabs.push_back(next_tab_id++); // Add new tab
    }

    // Submit our regular tabs
    for (int n = 0; n < active_tabs.Size; )
    {
      bool open = true;
      char name[16];
      snprintf(name, IM_ARRAYSIZE(name), "%04d", active_tabs[n]);

      if (ImGui::BeginTabItem(name, &open, ImGuiTabItemFlags_None))
      {
        ImGuiIO& io = ImGui::GetIO();
        if (io.ConfigFlags & ImGuiConfigFlags_DockingEnable)
        {
          std::string d_id = "MyDockSpace" + std::to_string(n);
          ImGuiID dockspace_id = ImGui::GetID(d_id.c_str());
          ImGui::DockSpace(dockspace_id, ImVec2(0.0f, 0.0f), dockspace_flags);
        }
        ImGui::EndTabItem();
      }
      else
      {
        std::string d_id = "MyDockSpace" + std::to_string(n);
        ImGuiID dockspace_id = ImGui::GetID(d_id.c_str());
        ImGui::DockSpace(dockspace_id, ImVec2(0.0f, 0.0f), ImGuiDockNodeFlags_KeepAliveOnly);
      }

      if (!open)
        active_tabs.erase(active_tabs.Data + n);
      else
        n++;
    }

    ImGui::EndTabBar();
  }
```

**Comment by PathogenDavid** [comment]
Thank you for filling out the whole issue template and doing your due diligence!

This is happening because the full ID of your dockspace is different depending on whether the tab item has focus. Dear ImGui IDs are stack-based, you can read more about this [in this FAQ entry](https://github.com/ocornut/imgui/blob/4fab72b40e2972f5d56fc884ed5797446238844e/docs/FAQ.md#q-about-the-id-stack-system).

Basically when tab `0000` has focus, your dockspace has the full ID of `[Dear ImGui Demo][MyTabBar][0000][MyDockSpace0]`. When it loses focus, you're calling `GetID` in a different ID scope resulting in the full ID of `[Dear ImGui Demo][MyTabBar][MyDockSpace0]`. (If you inspect the raw value of `dockspace_id` in each branch you'll find that they're different.

--------------

The easiest solution in your case would be to calculate `dockspace_id` outside of the `BeginTabItem` branch (which by extension is outside of the tab item's ID scope.) IE:

```cpp
std::string d_id = "MyDockSpace" + std::to_string(n);
ImGuiID dockspace_id = ImGui::GetID(d_id.c_str());

if (ImGui::BeginTabItem(name, &open, ImGuiTabItemFlags_None))
{
  ImGuiIO& io = ImGui::GetIO();
  if (io.ConfigFlags & ImGuiConfigFlags_DockingEnable)
  {
    ImGui::DockSpace(dockspace_id, ImVec2(0.0f, 0.0f), dockspace_flags);
  }
  ImGui::EndTabItem();
}
else
{
  ImGui::DockSpace(dockspace_id, ImVec2(0.0f, 0.0f), ImGuiDockNodeFlags_KeepAliveOnly);
}
```

----------

Some alternate solutions would be:

* Pass `ImGuiTabItemFlags_NoPushId` to `BeginTabItem` (could result in ID collisions between tabs if you ever add more than just the dockspace -- Use `PushID`/`PopID` to explicitly create a scope.)
* Manually call `PushID(name)`...`PopID()` around the `KeepAliveOnly` dockspace.
* Manually calculate a global ID using `ImHashStr` (an internal API)

**Comment by hka** [comment]
Ah! I see! 

I had just narrowed it down to being related to the dockspace_id, but didn't understand why. I change so I only used GetID once and then stored the value for each tab and saved the value, which solved the problem. With your explanation I also understand why :) 

If anyone else happens upon this, here is my working code:
https://github.com/hka/test-imgui/blob/9d615810079d0d2457a4c02661ab44d1e70ed599/src/simple.cpp

Thanks for the help!
